### DISCUSSION TITLE
Image magnification/level

### DISCUSSION THREAD

**Discussion by mhoibo** [FAST-Imaging/FAST#216] [Q&A] [answered]
Hi! 
I have had some issues with the patch generator in FAST. Previously, I have only used the level parameter to choose desired levels in images (I have mostly worked with images at x40) for patch generation, tissue microarray extraction etc.. But now I wanted to use the magnification parameter, since I might work with images scanned at different magnifications, and thus have different pyramid levels. But I have seen that with the images I have that are scanned at x40 (and then converted to ome.tiff, could the issue be here?), I get different results with level= 0 and magnification=40. I also get a warning that 
`Requested magnification level does not exist in image pyramid. Will now try to sample from a lower level and resize. This may increase runtime`
even though it is an image scanned at x40. I was wondering if anyone has tips on what the issue could be, and which parameter is best to use to be sure I get the magnification/level I want?
Also, how does FAST know which magnification levels exist in an image (if I use the magnification parameter, how is it determined if the magnification level is in the image or not?)
Thank you in advance!

**Comment by smistad** [comment]
We use the assumption that 40x means a pixel spacing of 0.00025 mm. Since a level might not have this exact spacing we need to use a threshold that might be too strict in your case. Here is the logic: https://github.com/smistad/FAST/blob/master/source/FAST/Algorithms/ImagePatch/PatchGenerator.cpp#L88
This logic can probably be improved

This off course requires the WSI file to have the correct spacing stored. Tiff files have a x and y resolution parameter.

What is the spacing/resolution of your file?

**Reply by mhoibo** [reply]
Thank you so much for the quick reply!

```
import fast
importer = fast.TIFFImagePyramidImporter.create(/path/to/image)
image = importer.runAndGetOutputData()
image.getSpacing()
```
yielded:
```
Level 0 has size 94172 149501 and tile size: 1024 1024 channels 3
Found 10 subdirectories....
Level 1 ...
...
Level 10 ...
Spacing from TIFF was 0.000172414 0.000172415Spacing:  [[1.7241413e-04]
[1.7241473e-04]
[1.0000000e+00]]
```
When checking another image I have that is scanned at x40, I get spacing 0.000136837
and some are also slightly different than these. 
I have used two different scanners, and some are also scanned with extended focal imaging (EFI), others not (as I remember most from one scanner are with EFI, and the others without EFI, but I can try to find that out if it helps with anything), do you think that could be the reason for the different spacing (around 0.00013 vs. around 0.00017)?

Do you think there is something wrong with the spacing in my ome.tiff files, that the wrong spacing is stored in the WSI files, or just that they actually have a different spacing?

Thank you again!

**Reply by smistad** [reply, marked as answer]
Hi @mhoibo 

I have been thinking more about this and note that my knowledge of microscopes is limited.
However resolution (e.g. pixel spacing) is not the same as magnification, although they are somewhat linked.
Check out this article on this exact topic: "Relationship between magnification and resolution in digital pathology systems" https://www.sciencedirect.com/science/article/pii/S2153353922006423

Thus, this gets rather complicated.. as using the spacing to "guess" the magnification of the images is not completely reliable.
We could try to get the magnification from the WSI metadata and use that if it exists as this is the most reliable source. If that doesn't exist we can try to guess from the pixel spacing.

**Reply by mhoibo** [reply]
Thank you, I will have a look at that paper. 

Yes, that sounds like a good idea!
